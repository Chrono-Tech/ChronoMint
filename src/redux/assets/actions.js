import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const ASSETS_LIST = 'assets/LIST'

export const listAssets = () => async (dispatch) => {
  const dao = await contractsManagerDAO.getAssetsManagerDAO()
  dispatch({type: ASSETS_LIST, list: await dao.getList()})
}
