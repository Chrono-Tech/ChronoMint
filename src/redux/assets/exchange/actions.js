import type ExchangeManagerDAO from 'dao/ExchangeManagerDAO'
import type ExchangeModel from 'models/ExchangeModel'
import type TokenModel from 'models/TokenModel'
import contractsManagerDAO from 'dao/ContractsManagerDAO'

export const EXCHANGE_MANAGER_LIST = 'exchangeManager/LIST'
export const EXCHANGE_MANAGER_SET = 'exchangeManager/SET'
export const EXCHANGE_MANAGER_REMOVE = 'exchangeManager/REMOVE'

const getDAO = async (): Promise<ExchangeManagerDAO> => await contractsManagerDAO.getExchangeManagerDAO()

export const listExchanges = () => async (dispatch) => {
  const dao = await getDAO()
  dispatch({type: EXCHANGE_MANAGER_LIST, list: await dao.getList()})
}

export const addExchange = (address) => async () => {
  const exchange = new ExchangeModel({address, isFetching: true})
  dispatch({type: EXCHANGE_MANAGER_SET, exchange})
  try {
    const dao = await getDAO()
    await dao.addExchange(address)
  } catch (e) {
    // no rollback
  }
}

export const createExchange = (token: TokenModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.createExchange(token)
  } catch (e) {
    // no rollback
  }
}

export const editExchange = (exchange: ExchangeModel, newExchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.editExchange(exchange, newExchange)
  } catch (e) {
    // no rollback
  }
}

export const removeExchange = (exchange: ExchangeModel) => async () => {
  try {
    const dao = await getDAO()
    await dao.removeExchange(exchange)
  } catch (e) {
    // no rollback
  }
}

// TODO @bshevchenko: MINT-230 AssetsManager
// export const addOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.addOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
//
// export const removeOwner = (exchange: ExchangeModel, account) => async () => {
//   try {
//     const dao = await getDAO()
//     await dao.removeOwner(exchange, account)
//   } catch (e) {
//     // no rollback
//   }
// }
