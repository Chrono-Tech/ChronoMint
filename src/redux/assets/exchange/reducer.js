import Immutable from 'immutable'
import * as a from './actions'

const initialState = {
  list: new Immutable.Map(),
  isFetched: false
}

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case a.EXCHANGE_MANAGER_LIST:
      return {
        ...state,
        list: action.list,
        isFetched: true
      }
    default:
      return state
  }
}

export default reducer
