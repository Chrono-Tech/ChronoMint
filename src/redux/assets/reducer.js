import Immutable from 'immutable'
import * as a from './actions'

const initialState = {
  list: new Immutable.Map(),
  isFetched: false
}

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case a.ASSETS_LIST:
      return {
        ...state,
        list: action.list,
        isFetched: true
      }
    default:
      return state
  }
}

export default reducer
