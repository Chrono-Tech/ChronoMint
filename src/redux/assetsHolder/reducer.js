import TimeHolderModel from 'models/timeHolder/TimeHolderModel'
import * as a from './actions'

const initialState = new TimeHolderModel()

export default (state = initialState, action) => {
  switch (action.type) {
    case a.TIME_HOLDER_INIT:
      return state.isInited(true)
    case a.TIME_HOLDER_TIME_ADDRESS:
      return state.timeAddress(action.address)
    case a.TIME_HOLDER_ADDRESS:
      return state.timeHolderAddress(action.address)
    case a.TIME_HOLDER_WALLET_ADDRESS:
      return state.timeHolderWalletAddress(action.address)
    default:
      return state
  }
}
