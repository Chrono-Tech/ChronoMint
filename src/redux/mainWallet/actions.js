import BigNumber from 'bignumber.js'
import { TXS_PER_PAGE } from 'dao/AbstractTokenDAO'
import assetDonatorDAO from 'dao/AssetDonatorDAO'
import contractsManagerDAO from 'dao/ContractsManagerDAO'
import ethereumDAO from 'dao/EthereumDAO'
import Immutable from 'immutable'
import { bccProvider, btcProvider } from 'Login/network/BitcoinProvider'
import { nemProvider } from 'Login/network/NemProvider'
import ApprovalNoticeModel from 'models/notices/ApprovalNoticeModel'
import TransferNoticeModel from 'models/notices/TransferNoticeModel'
import type ProfileModel from 'models/ProfileModel'
import TokenModel from 'models/TokenModel'
import type TxModel from 'models/TxModel'
import { notify } from 'redux/notifier/actions'
import { DUCK_SESSION } from 'redux/session/actions'
import tokenService, { EVENT_NEW_TOKEN } from 'services/TokenService'
import { addMarketToken } from '../market/action'

export const DUCK_MAIN_WALLET = 'mainWallet'

export const WALLET_TOKENS_FETCH = 'mainWallet/TOKENS_FETCH'
export const WALLET_TOKENS = 'mainWallet/TOKENS'
export const WALLET_BALANCE = 'mainWallet/BALANCE'
export const WALLET_BALANCE_SET = 'mainWallet/BALANCE_SET'
export const WALLET_ALLOWANCE = 'mainWallet/ALLOWANCE'
export const WALLET_TIME_DEPOSIT = 'mainWallet/TIME_DEPOSIT'
export const WALLET_TIME_ADDRESS = 'mainWallet/TIME_ADDRESS'
export const WALLET_BTC_ADDRESS = 'mainWallet/BTC_ADDRESS'
export const WALLET_BCC_ADDRESS = 'mainWallet/BCC_ADDRESS'
export const WALLET_NEM_ADDRESS = 'mainWallet/NEM_ADDRESS'
export const WALLET_TRANSACTIONS_FETCH = 'mainWallet/TRANSACTIONS_FETCH'
export const WALLET_TRANSACTION = 'mainWallet/TRANSACTION'
export const WALLET_TRANSACTIONS = 'mainWallet/TRANSACTIONS'
export const WALLET_IS_TIME_REQUIRED = 'mainWallet/IS_TIME_REQUIRED'

export const ETH = ethereumDAO.getSymbol()
export const TIME = 'TIME'
export const LHT = 'LHT'

export const setBalance = (token: TokenModel, amount: BigNumber) => ({
  type: WALLET_BALANCE_SET,
  token,
  amount,
})

export const updateBalance = (token: TokenModel, isCredited, amount: BigNumber) => ({
  type: WALLET_BALANCE,
  token,
  isCredited,
  amount,
})

export const balancePlus = (amount: BigNumber, token: TokenModel) => updateBalance(token, true, amount)

export const balanceMinus = (amount: BigNumber, token: TokenModel) => updateBalance(token, false, amount)

export const updateDeposit = (amount: BigNumber, isCredited: ?boolean) => ({
  type: WALLET_TIME_DEPOSIT,
  isCredited,
  amount,
})

export const depositPlus = (amount: BigNumber) => updateDeposit(amount, true)

export const depositMinus = (amount: BigNumber) => updateDeposit(amount, false)

export const allowance = (token: TokenModel, value: BigNumber, spender) => ({
  type: WALLET_ALLOWANCE,
  token,
  value,
  spender,
})

export const watchTransfer = (notice: TransferNoticeModel) => async (dispatch, getState) => {
  const tx: TxModel = notice.tx()
  const token: TokenModel = getState().get(DUCK_MAIN_WALLET).tokens().get(tx.symbol())

  dispatch(updateBalance(token, tx.isCredited(), tx.value()))

  const timeHolderDAO = await contractsManagerDAO.getTIMEHolderDAO()
  const timeHolderWalletAddress = await timeHolderDAO.getWalletAddress()
  let updateTIMEAllowance = false
  if (tx.to() === timeHolderWalletAddress) {
    dispatch(depositPlus(tx.value()))
    updateTIMEAllowance = true
  } else if (tx.from() === timeHolderWalletAddress) {
    dispatch(depositMinus(tx.value()))
    updateTIMEAllowance = true
  }
  if (updateTIMEAllowance) {
    const dao = await token.dao()
    dispatch(allowance(token, await dao.getAccountAllowance(timeHolderWalletAddress), timeHolderWalletAddress))
  }

  dispatch(notify(notice))
  dispatch({ type: WALLET_TRANSACTION, tx })
}

export const watchBalance = ({ symbol, balance /* balance3, balance6 */ }) => async (dispatch, getState) => {
  const token: TokenModel = getState().get(DUCK_MAIN_WALLET).tokens().get(symbol)
  dispatch(setBalance(token, balance))
}

export const watchInitWallet2 = () => (dispatch, getState) => {
  tokenService.on(EVENT_NEW_TOKEN, (token: TokenModel) => {
    const { profile } = getState().get(DUCK_SESSION)

    console.log('--actions4444#', token.toJS(), profile.toJS())
  })
}

export const watchInitWallet = () => async (dispatch, getState) => {
  const state = getState()

  const profile: ProfileModel = state.get(DUCK_SESSION).profile
  const previous = state.get(DUCK_MAIN_WALLET).tokens()

  dispatch({ type: WALLET_TOKENS_FETCH })
  const dao = await contractsManagerDAO.getERC20ManagerDAO()
  let tokens = await dao.getUserTokens(profile.tokens().toArray())
  dispatch({ type: WALLET_TOKENS, tokens })
  dispatch(getAccountTransactions(tokens))

  const toStopArray = previous.filter((k) => !tokens.get(k)).valueSeq().toArray().map((token: TokenModel) => {
    const dao = token.dao()
    return dao.stopWatching()
  })
  if (toStopArray.length) {
    await Promise.all(toStopArray)
  }

  const timeHolderDAO = await contractsManagerDAO.getTIMEHolderDAO()
  const [ timeHolderAddress, timeHolderWalletAddress ] = await Promise.all([
    timeHolderDAO.getAddress(),
    timeHolderDAO.getWalletAddress(),
  ])

  let contractNames = {}
  contractNames[ timeHolderAddress ] = TIME + ' Holder'
  ApprovalNoticeModel.setContractNames(contractNames)
  dispatch({ type: WALLET_TIME_ADDRESS, address: timeHolderWalletAddress })

  // NOTE @ipavlenko: BCC and BTC addresses usually the same.
  // Decided to manage them independently to simplify further works on multiple wallets. .
  dispatch({ type: WALLET_BTC_ADDRESS, address: btcProvider.getAddress() })
  dispatch({ type: WALLET_BCC_ADDRESS, address: bccProvider.getAddress() })
  dispatch({ type: WALLET_NEM_ADDRESS, address: nemProvider.getAddress() })

  tokens = tokens.filter((k) => !previous.get(k)).valueSeq().toArray()
  for (let token: TokenModel of tokens) {
    dispatch(addMarketToken(token.symbol()))
    const dao = token.dao()
    await dao.watchTransfer((notice) => dispatch(watchTransfer(notice)))
    if (dao.watchBalance) {
      await dao.watchBalance((balance) => dispatch(watchBalance(balance)))
    }
    await dao.watchApproval((notice: ApprovalNoticeModel) => {
      dispatch({ type: WALLET_ALLOWANCE, token, value: notice.value(), spender: notice.spender() })
      dispatch(notify(notice.setToken(token)))
    })
  }
}

export const mainTransfer = (token: TokenModel, amount: string, recipient) => async (dispatch) => {
  const amountLocal = new BigNumber(amount)

  dispatch(balanceMinus(amountLocal, token))
  // TODO @bshevchenko: sub balances with values of outcome pending transactions
  try {
    const dao = await token.dao()
    await dao.transfer(recipient, amountLocal)
  } finally {
    // compensation for update in watchTransfer
    dispatch(balancePlus(amountLocal, token))
  }
}

export const mainApprove = (token: TokenModel, amount: Number, spender: string) => async () => {
  try {
    const amountBN = new BigNumber(amount)
    const dao = await token.dao()
    await dao.approve(spender, amountBN)
  } catch (e) {
    // no rollback
    // eslint-disable-next-line
    console.error('approve error', e.message)
  }
}

export const depositTIME = (amount: string) => async (dispatch, getState) => {
  const amountBN = new BigNumber(amount)
  const token: TokenModel = getState().get(DUCK_MAIN_WALLET).tokens().get(TIME)

  dispatch(balanceMinus(amountBN, token))

  try {
    const dao = await contractsManagerDAO.getTIMEHolderDAO()
    await dao.deposit(amountBN)
  } finally {
    // compensation for update in watchTransfer
    dispatch(balancePlus(amountBN, token))
  }
}

export const withdrawTIME = (amount: string) => async (dispatch) => {
  const amountBN = new BigNumber(amount)

  dispatch(depositMinus(amountBN))

  try {
    const dao = await contractsManagerDAO.getTIMEHolderDAO()
    await dao.withdraw(amountBN)
  } finally {
    dispatch(depositPlus(amountBN))
  }
}

export const initTIMEDeposit = () => async (dispatch) => {
  const dao = await contractsManagerDAO.getTIMEHolderDAO()
  const deposit = await dao.getAccountDepositBalance()
  dispatch(updateDeposit(deposit, null))
}

export const updateIsTIMERequired = () => async (dispatch) => {
  dispatch({ type: WALLET_IS_TIME_REQUIRED, value: await assetDonatorDAO.isTIMERequired() })
}

export const requireTIME = () => async (dispatch) => {
  try {
    await assetDonatorDAO.requireTIME()
  } catch (e) {
    // no rollback
    // eslint-disable-next-line
    console.error('require time error', e.message)
  }
  await dispatch(updateIsTIMERequired())
}

/**
 * LATEST TRANSACTIONS
 */
const getTransferId = 'wallet'
let lastCacheId
let txsCache = []

export const getAccountTransactions = (tokens) => async (dispatch) => {
  dispatch({ type: WALLET_TRANSACTIONS_FETCH })

  const tokensLocal = tokens.valueSeq().toArray()

  const cacheId = Object.values(tokensLocal).map((v: TokenModel) => v.symbol()).join(',')

  const reset = lastCacheId && cacheId !== lastCacheId
  lastCacheId = cacheId
  if (reset) {
    txsCache = []
  }

  let txs = txsCache.slice(0, TXS_PER_PAGE)
  txsCache = txsCache.slice(TXS_PER_PAGE)

  if (txs.length < TXS_PER_PAGE) { // so cache is empty
    const promises = []
    for (let token: TokenModel of tokensLocal) {
      if (reset) {
        token.dao().resetFilterCache()
      }
      promises.push(token.dao().getTransfer(getTransferId))
    }
    const result = await Promise.all(promises)

    let newTxs = []
    for (let pack of result) {
      newTxs = [ ...newTxs, ...pack ]
    }

    newTxs.sort((a, b) => b.get('time') - a.get('time'))

    txs = [ ...txs, ...newTxs ]
    txsCache = txs.slice(TXS_PER_PAGE)
    txs = txs.slice(0, TXS_PER_PAGE)
  }

  let map = new Immutable.Map()
  for (let tx: TxModel of txs) {
    map = map.set(tx.id(), tx)
  }

  dispatch({ type: WALLET_TRANSACTIONS, map })
}
