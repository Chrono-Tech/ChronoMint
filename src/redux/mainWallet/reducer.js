import MainWalletModel from 'models/Wallet/MainWalletModel'
import * as a from './actions'

const initialState = new MainWalletModel()

export default (state = initialState, action) => {
  switch (action.type) {
    case a.WALLET_TOKENS_FETCH:
      return state.isFetching(true)
    case a.WALLET_TOKENS:
      return state
        .tokens(action.tokens)
        .isFetching(false)
        .isFetched(true)
    case a.WALLET_BALANCE:
      return state.tokens(state.tokens().set(
        action.token.id(),
        state.tokens().get(action.token.id()).updateBalance(action.isCredited, action.amount)
      ))
    case a.WALLET_BALANCE_SET:
      return state.tokens(state.tokens().set(
        action.token.id(),
        state.tokens().get(action.token.id()).setBalance(action.amount)
      ))
    case a.WALLET_ALLOWANCE:
      return state.tokens(state.tokens().set(
        action.token.id(),
        state.tokens().get(action.token.id()).setAllowance(action.spender, action.value)
      ))
    case a.WALLET_TIME_DEPOSIT:
      // TODO @dkchv: !!! action.isCredited
      return state.timeDeposit(action.amount)
    case a.WALLET_TIME_ADDRESS:
      return state.timeAddress(action.address)
    case a.WALLET_BTC_ADDRESS:
      return state.btcAddress(action.address)
    case a.WALLET_BCC_ADDRESS:
      return state.bccAddress(action.address)
    case a.WALLET_TRANSACTIONS_FETCH:
      return state.transactions(state.transactions().isFetching(true))
    case a.WALLET_TRANSACTION:
      return state.transactions(state.transactions().list(
        state.transactions().list().set(action.tx.id(), action.tx)
      ))
    case a.WALLET_TRANSACTIONS:
      return state.transactions(state.transactions()
        .isFetching(false)
        .list(state.transactions().list().merge(action.map))
        .endOfList(action.map.size)
      )
    case a.WALLET_IS_TIME_REQUIRED:
      return state.isTIMERequired(action.value)
    default:
      return state
  }
}
