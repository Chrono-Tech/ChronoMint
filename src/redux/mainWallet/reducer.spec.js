import BigNumber from 'bignumber.js'
import Immutable from 'immutable'
import { accounts } from 'specsInit'
import MainWallet from 'models/Wallet/MainWalletModel'
import TokenModel from 'models/TokenModel'
import TxModel from 'models/TxModel'
import * as a from './actions'
import reducer from './reducer'

const token1 = new TokenModel({ symbol: 'TK1' })
const token2 = new TokenModel({ symbol: 'TK2' })

const tokens = new Immutable.Map({ [token1.symbol()]: token1 })

const tx1 = new TxModel({ txHash: 'hash1', from: 1, to: 2 })
const tx2 = new TxModel({ txHash: 'hash2', from: 3, to: 4 })

describe('settings wallet reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toEqual(new MainWallet())
  })

  it('should handle WALLET_TOKENS_FETCH', () => {
    expect(reducer({ tokensFetching: false }, { type: a.WALLET_TOKENS_FETCH })).toEqual({
      tokensFetching: true,
    })
  })

  it('should handle WALLET_TOKENS', () => {
    const tokens = new Immutable.Map({
      [token1.symbol()]: token1,
      [token2.symbol()]: token2,
    })
    expect(reducer({}, { type: a.WALLET_TOKENS, tokens })).toEqual({
      tokensFetching: false,
      tokensFetched: true,
      tokens,
    })
  })

  it('should handle WALLET_BALANCE', () => {
    expect(reducer({
      tokens,
    }, {
      type: a.WALLET_BALANCE, token: token1, isCredited: true, amount: 5,
    })).toEqual({
      tokens: new Immutable.Map({
        [token1.symbol()]: token1.updateBalance(true, 5),
      }),
    })
  })

  it('should handle WALLET_ALLOWANCE', () => {
    expect(reducer({
      tokens,
    }, {
      type: a.WALLET_ALLOWANCE, token: token1, spender: accounts[4], value: 4,
    })).toEqual({
      tokens: new Immutable.Map({
        [token1.symbol()]: token1.setAllowance(accounts[4], 4),
      }),
    })
  })

  it('should handle WALLET_TIME_DEPOSIT', () => {
    expect(reducer({
      timeDeposit: new BigNumber(5),
    }, {
      type: a.WALLET_TIME_DEPOSIT, isCredited: false, amount: 3,
    })).toEqual({
      timeDeposit: new BigNumber(2),
    })
  })

  it('should handle WALLET_TIME_DEPOSIT', () => {
    expect(reducer([], { type: a.WALLET_TIME_ADDRESS, address: accounts[5] })).toEqual({
      timeAddress: accounts[5],
    })
  })

  it('should handle WALLET_TRANSACTIONS_FETCH', () => {
    const initial = {
      transactions: {
        list: new Immutable.Map({ a: 1 }),
        isFetching: false,
      },
    }
    expect(reducer(initial, { type: a.WALLET_TRANSACTIONS_FETCH })).toEqual({
      transactions: {
        list: new Immutable.Map({ a: 1 }),
        isFetching: true,
      },
    })
  })

  it('should handle WALLET_TRANSACTION', () => {
    const initial = {
      transactions: {
        list: new Immutable.Map({
          [tx1.id()]: tx1,
        }),
      },
    }
    const updatedTx = new TxModel({
      txHash: 'hash1', from: 1, to: 2, blockNumber: 10,
    })

    expect(reducer(initial, { type: a.WALLET_TRANSACTION, tx: updatedTx })).toEqual({
      transactions: {
        list: new Immutable.Map({
          'hash1 - 1 - 2': updatedTx,
        }),
      },
    })
  })

  it('should handle WALLET_TRANSACTIONS', () => {
    const initial = {
      transactions: {
        list: new Immutable.Map({
          tx1,
        }),
        endOfList: true,
        isFetching: true,
      },
    }
    expect(reducer(initial, { type: a.WALLET_TRANSACTIONS, map: { tx2 } })).toEqual({
      transactions: {
        list: new Immutable.Map({ tx1, tx2 }),
        endOfList: false,
        isFetching: false,
      },
    })
  })

  it('should handle WALLET_IS_TIME_REQUIRED', () => {
    expect(reducer({ isTIMERequired: true }, { type: a.WALLET_IS_TIME_REQUIRED, value: false })).toEqual({
      isTIMERequired: false,
    })
  })
})
