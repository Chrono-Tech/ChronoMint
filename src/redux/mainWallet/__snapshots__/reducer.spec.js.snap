// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings wallet reducer should handle WALLET_ALLOWANCE 1`] = `
Object {
  "address": null,
  "bccAddress": null,
  "btcAddress": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isTIMERequired": true,
  "nemAddress": null,
  "timeAddress": null,
  "timeDeposit": "0",
  "timestamp": null,
  "tokens": Object {
    "TK1": Object {
      "address": null,
      "allowance": Object {
        "0xee42d0ea34b0b40d1930ed05ed2845d08ea7cfd1": 4,
      },
      "balance": "0",
      "dao": null,
      "decimals": null,
      "fee": null,
      "feeAddress": null,
      "icon": null,
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isOptional": true,
      "isPending": false,
      "isReissuable": null,
      "managersList": null,
      "name": null,
      "platform": null,
      "symbol": "TK1",
      "timestamp": null,
      "totalSupply": "0",
      "transactionHash": null,
      "url": null,
      "withFee": null,
    },
  },
  "transactionHash": null,
  "transactions": Object {
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isPending": false,
    "list": Object {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null,
  },
}
`;

exports[`settings wallet reducer should handle WALLET_BALANCE 1`] = `
Object {
  "address": null,
  "bccAddress": null,
  "btcAddress": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isTIMERequired": true,
  "nemAddress": null,
  "timeAddress": null,
  "timeDeposit": "0",
  "timestamp": null,
  "tokens": Object {
    "TK1": Object {
      "address": null,
      "allowance": Object {},
      "balance": "5",
      "dao": null,
      "decimals": null,
      "fee": null,
      "feeAddress": null,
      "icon": null,
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isOptional": true,
      "isPending": false,
      "isReissuable": null,
      "managersList": null,
      "name": null,
      "platform": null,
      "symbol": "TK1",
      "timestamp": null,
      "totalSupply": "0",
      "transactionHash": null,
      "url": null,
      "withFee": null,
    },
  },
  "transactionHash": null,
  "transactions": Object {
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isPending": false,
    "list": Object {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null,
  },
}
`;

exports[`settings wallet reducer should handle WALLET_IS_TIME_REQUIRED 1`] = `
Object {
  "address": null,
  "bccAddress": null,
  "btcAddress": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isTIMERequired": false,
  "nemAddress": null,
  "timeAddress": null,
  "timeDeposit": "0",
  "timestamp": null,
  "tokens": Object {},
  "transactionHash": null,
  "transactions": Object {
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isPending": false,
    "list": Object {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null,
  },
}
`;

exports[`settings wallet reducer should handle WALLET_TIME_DEPOSIT 1`] = `
Object {
  "address": null,
  "bccAddress": null,
  "btcAddress": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isTIMERequired": true,
  "nemAddress": null,
  "timeAddress": null,
  "timeDeposit": 3,
  "timestamp": null,
  "tokens": Object {},
  "transactionHash": null,
  "transactions": Object {
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isPending": false,
    "list": Object {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null,
  },
}
`;

exports[`settings wallet reducer should handle WALLET_TIME_DEPOSIT 2`] = `
Object {
  "address": null,
  "bccAddress": null,
  "btcAddress": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isTIMERequired": true,
  "nemAddress": null,
  "timeAddress": "0xec8a163d2ebc655fd44a7b159ca637abb380339c",
  "timeDeposit": "0",
  "timestamp": null,
  "tokens": Object {},
  "transactionHash": null,
  "transactions": Object {
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isPending": false,
    "list": Object {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null,
  },
}
`;

exports[`settings wallet reducer should handle WALLET_TOKENS 1`] = `
Object {
  "address": null,
  "bccAddress": null,
  "btcAddress": null,
  "id": null,
  "isFailed": false,
  "isFetched": true,
  "isFetching": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isTIMERequired": true,
  "nemAddress": null,
  "timeAddress": null,
  "timeDeposit": "0",
  "timestamp": null,
  "tokens": Object {
    "TK1": Object {
      "address": null,
      "allowance": Object {},
      "balance": "0",
      "dao": null,
      "decimals": null,
      "fee": null,
      "feeAddress": null,
      "icon": null,
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isOptional": true,
      "isPending": false,
      "isReissuable": null,
      "managersList": null,
      "name": null,
      "platform": null,
      "symbol": "TK1",
      "timestamp": null,
      "totalSupply": "0",
      "transactionHash": null,
      "url": null,
      "withFee": null,
    },
    "TK2": Object {
      "address": null,
      "allowance": Object {},
      "balance": "0",
      "dao": null,
      "decimals": null,
      "fee": null,
      "feeAddress": null,
      "icon": null,
      "id": null,
      "isFailed": false,
      "isFetched": false,
      "isFetching": false,
      "isOptional": true,
      "isPending": false,
      "isReissuable": null,
      "managersList": null,
      "name": null,
      "platform": null,
      "symbol": "TK2",
      "timestamp": null,
      "totalSupply": "0",
      "transactionHash": null,
      "url": null,
      "withFee": null,
    },
  },
  "transactionHash": null,
  "transactions": Object {
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isPending": false,
    "list": Object {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null,
  },
}
`;

exports[`settings wallet reducer should handle WALLET_TOKENS_FETCH 1`] = `
Object {
  "address": null,
  "bccAddress": null,
  "btcAddress": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": true,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isTIMERequired": true,
  "nemAddress": null,
  "timeAddress": null,
  "timeDeposit": "0",
  "timestamp": null,
  "tokens": Object {},
  "transactionHash": null,
  "transactions": Object {
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isPending": false,
    "list": Object {},
    "selected": null,
    "timestamp": null,
    "transactionHash": null,
  },
}
`;

exports[`settings wallet reducer should handle WALLET_TRANSACTION 1`] = `
Object {
  "address": null,
  "bccAddress": null,
  "btcAddress": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isTIMERequired": true,
  "nemAddress": null,
  "timeAddress": null,
  "timeDeposit": "0",
  "timestamp": null,
  "tokens": Object {},
  "transactionHash": null,
  "transactions": Object {
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isPending": false,
    "list": Object {
      "hash1 - 1 - 2": Object {
        "args": null,
        "blockHash": null,
        "blockNumber": 10,
        "by": null,
        "credited": null,
        "from": 1,
        "gas": null,
        "gasFee": "0",
        "gasPrice": null,
        "id": null,
        "input": null,
        "symbol": "",
        "time": null,
        "timestamp": null,
        "to": 2,
        "transactionIndex": null,
        "txHash": "hash1",
        "type": "",
        "value": "0",
      },
    },
    "selected": null,
    "timestamp": null,
    "transactionHash": null,
  },
}
`;

exports[`settings wallet reducer should handle WALLET_TRANSACTIONS 1`] = `
Object {
  "address": null,
  "bccAddress": null,
  "btcAddress": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isTIMERequired": true,
  "nemAddress": null,
  "timeAddress": null,
  "timeDeposit": "0",
  "timestamp": null,
  "tokens": Object {},
  "transactionHash": null,
  "transactions": Object {
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": false,
    "isPending": false,
    "list": Object {
      "tx1": Object {
        "args": null,
        "blockHash": null,
        "blockNumber": null,
        "by": null,
        "credited": null,
        "from": 1,
        "gas": null,
        "gasFee": "0",
        "gasPrice": null,
        "id": null,
        "input": null,
        "symbol": "",
        "time": null,
        "timestamp": null,
        "to": 2,
        "transactionIndex": null,
        "txHash": "hash1",
        "type": "",
        "value": "0",
      },
      "tx2": Object {
        "args": null,
        "blockHash": null,
        "blockNumber": null,
        "by": null,
        "credited": null,
        "from": 3,
        "gas": null,
        "gasFee": "0",
        "gasPrice": null,
        "id": null,
        "input": null,
        "symbol": "",
        "time": null,
        "timestamp": null,
        "to": 4,
        "transactionIndex": null,
        "txHash": "hash2",
        "type": "",
        "value": "0",
      },
    },
    "selected": null,
    "timestamp": null,
    "transactionHash": null,
  },
}
`;

exports[`settings wallet reducer should handle WALLET_TRANSACTIONS_FETCH 1`] = `
Object {
  "address": null,
  "bccAddress": null,
  "btcAddress": null,
  "id": null,
  "isFailed": false,
  "isFetched": false,
  "isFetching": false,
  "isMainWallet": true,
  "isMultisig": false,
  "isPending": false,
  "isTIMERequired": true,
  "nemAddress": null,
  "timeAddress": null,
  "timeDeposit": "0",
  "timestamp": null,
  "tokens": Object {},
  "transactionHash": null,
  "transactions": Object {
    "endOfList": false,
    "id": null,
    "isFailed": false,
    "isFetched": false,
    "isFetching": true,
    "isPending": false,
    "list": Object {
      "a": 1,
    },
    "selected": null,
    "timestamp": null,
    "transactionHash": null,
  },
}
`;
